File: py110/small_problems_110_119/medium2/5.py
# https://launchschool.com/exercises/07af1741

"""

input: integer
output: next featured integer or error
rules:
1. output has to be greater then input
2. featured number is odd
3. multiple of 7
4. all digits occcuring exactly once. 123


7 
14
21 
28

12 + 1 
13 + 1 


Algo:
1. 

2. Function to check if number is odd multiple of 7 and occurs exactly once
3. number is odd by % 2 != 0, number % 7, len(set(str)) == len(str)

"""

def check_if_valid_feature(num):
    num += 1
    while num % 7 != 0 or num % 2 == 0:
        num += 1
    
    return num

def check_unique(num):
    while len(set(str(num))) != len(str(num)):
        num += 14
    return num


def next_featured(integer):
    error = ("There is no possible number that "
         "fulfills those requirements.")

    num = integer

    while num < 9876543201:

     
        num = check_if_valid_feature(integer)
        unique_num = check_unique(num)
        return unique_num
    
    
    return error

# print(next_featured(12) == 21)                  # True
# print(next_featured(20) == 21)                  # True
# print(next_featured(21) == 35)                  # True
# print(next_featured(997) == 1029)               # True
# print(next_featured(1029) == 1043)              # True
# print(next_featured(999999) == 1023547)         # True
# print(next_featured(999999987) == 1023456987)   # True
print(next_featured(9876543186) == 9876543201)  # True
print(next_featured(9876543200) == 9876543201)  # True

error = ("There is no possible number that "
         "fulfills those requirements.")
print(next_featured(9876543201) == error)       # True
    
    
             

File: ls215/smallProblems/med2/5.js
// https://launchschool.com/exercises/c1bfff07

'use strict';

/*

input: Number
output: featured number. Else error message
rules:
1. Odd number that is a multiple of 7
2. With all of its digit occuring exactly once.

49 Yes

98 No not odd


133 No because 3 appears twice.




Q:

DS:


Algo:
0. If number is 9876543201 then return error message

1. See if the current number is a multiple of 7.
2. IF not then increment it till it is
  3. Check if it is even. IF it is even add 7
  3a. Check if it has dups. if it does add 14.
  4. IF it is not that is our number. and return number
3. Else add 14 and check if there are dups.
  if there are dups then add 14.

*/


function featured(number) {
  if (number === 9876543201) return 'There is no possible number that fulfills those requirements.';

  return findNumber(number);

}

function findNumber(num) {
  if (num % 7 === 0 && !noDups(num)) return num + 14;

  let nextNum = num += 1;

  while (nextNum % 2 === 0 || nextNum % 7 !== 0) {
    nextNum += 1;
  }

  while (noDups(nextNum)) {
    nextNum += 14;
  }

  return nextNum;
}

function noDups(num) {
  let strNum = String(num);
  let set = new Set(strNum);

  if (set.size !== strNum.length) return true;

  return false;
}


console.log(featured(12));           // 21
console.log(featured(20));           // 21
console.log(featured(21));           // 35
console.log(featured(997));          // 1029
console.log(featured(1029));         // 1043
console.log(featured(999999));       // 1023547
console.log(featured(999999987));    // 1023456987
console.log(featured(9876543186));   // 9876543201
console.log(featured(9876543200));   // 9876543201
console.log(featured(9876543201));   // "There is no possible number that fulfills those requirements."